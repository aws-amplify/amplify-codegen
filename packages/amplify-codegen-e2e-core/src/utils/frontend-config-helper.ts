export enum AmplifyFrontend {
  javascript = 'javascript',
  ios = 'ios',
  android = 'android',
  flutter = 'flutter'
}

export type AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend;
  srcDir: string;
  modelgenDir: string;
  graphqlCodegenDir?: string;
}

export const DEFAULT_JS_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.javascript,
  srcDir: 'src',
  modelgenDir: 'src/models',
  graphqlCodegenDir : 'src/graphql'
};

export const DEFAULT_ANDROID_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.android,
  srcDir: 'app/src/main/res',
  modelgenDir: 'app/src/main/java',
  graphqlCodegenDir : 'app/src/main/graphql'
};

export const DEFAULT_IOS_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.ios,
  srcDir: '.',
  modelgenDir: 'amplify/generated/models',
  graphqlCodegenDir : 'graphql'
};

export const DEFAULT_FLUTTER_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.flutter,
  srcDir: 'lib',
  modelgenDir: 'lib/models',
};