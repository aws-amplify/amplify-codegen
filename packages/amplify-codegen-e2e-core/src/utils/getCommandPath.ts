import { execSync } from 'child_process';
/* Get the full absolute path of a command with `which`
 *
 * Windows executors on CB fail to spawn a command with custom nexpect unless the full path is used.
 */
export function getCommandPath(command: string): string {
  return execSync(`which ${command}`)
    .toString()
    .trim();
}
