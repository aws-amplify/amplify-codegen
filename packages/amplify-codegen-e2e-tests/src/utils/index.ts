import { existsSync, writeFileSync, readdirSync, mkdirSync } from "fs";
import path from 'path';

export function isNotEmptyDir(dirPath: string) : boolean {
  return existsSync(dirPath) && readdirSync(dirPath).length > 0;
}

export function generateSourceCode(projectRoot: string, srcDir: string) : string {
  const userFileData = 'This is a pre-existing file.';
  const srcCodePath = path.join(projectRoot, srcDir, 'sample.txt');
  if (!existsSync(path.dirname(srcCodePath))) {
    mkdirSync(path.dirname(srcCodePath), {recursive: true});
  }
  writeFileSync(srcCodePath, userFileData);
  return srcCodePath;
}