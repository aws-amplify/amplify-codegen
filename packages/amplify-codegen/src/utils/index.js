const getFrontEndHandler = require('./getFrontEndHandler');
const getFrontEndFramework = require('./getFrontEndFramework');
const getAppSyncAPIDetails = require('./getAppSyncAPIDetails');
const downloadIntrospectionSchema = require('./downloadIntrospectionSchema');
const getSchemaDownloadLocation = require('./getSchemaDownloadLocation');
const getIncludePattern = require('./getIncludePattern');
const getAppSyncAPIInfo = require('./getAppSyncAPIInfo');
const getAppSyncAPIInfoFromProject = require('./getAppSyncAPIInfoFromProject');
const getProjectAwsRegion = require('./getProjectAWSRegion');
const getGraphQLDocPath = require('./getGraphQLDocPath');
const downloadIntrospectionSchemaWithProgress = require('./generateIntrospectionSchemaWithProgress');
const isAppSyncApiPendingPush = require('./isAppSyncApiPendingPush');
const updateAmplifyMeta = require('./updateAmplifyMeta');
const isCodegenConfigured = require('./isCodegenConfigured');
const getSDLSchemaLocation = require('./getSDLSchemaLocation');
const switchToSDLSchema = require('./switchToSDLSchema');
const ensureIntrospectionSchema = require('./ensureIntrospectionSchema');
const { readSchemaFromFile } = require('./readSchemaFromFile');
const defaultDirectiveDefinitions = require('./defaultDirectiveDefinitions');
module.exports = {
  getAppSyncAPIDetails,
  getFrontEndHandler,
  getFrontEndFramework,
  getSchemaDownloadLocation,
  downloadIntrospectionSchema,
  downloadIntrospectionSchemaWithProgress,
  getIncludePattern,
  getAppSyncAPIInfo,
  getAppSyncAPIInfoFromProject,
  getProjectAwsRegion,
  getGraphQLDocPath,
  isAppSyncApiPendingPush,
  updateAmplifyMeta,
  isCodegenConfigured,
  getSDLSchemaLocation,
  switchToSDLSchema,
  ensureIntrospectionSchema,
  readSchemaFromFile,
  defaultDirectiveDefinitions,
};
