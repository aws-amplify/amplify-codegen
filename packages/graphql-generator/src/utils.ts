import { Language, Platform, Target, FileExtension } from './type';

export const targetToFileExtensionMap: { [key in Target]: FileExtension } = {
  javascript: 'js',
  graphql: 'graphql',
  flow: 'js',
  typescript: 'ts',
  angular: 'graphql',
};

export const platformToLanguageMap: { [key in Platform]: Language } = {
  android: 'java',
  ios: 'swift',
  flutter: 'dart',
  javascript: 'javascript',
  introspection: 'introspection',
};
