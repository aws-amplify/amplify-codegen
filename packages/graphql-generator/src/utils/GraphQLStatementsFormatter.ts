import prettier, { BuiltInParserName } from 'prettier';
import {
  interfaceNameFromOperation,
  interfaceVariablesNameFromOperation,
} from '@aws-amplify/graphql-types-generator/lib/typescript/codeGeneration';
import type { GraphQLWithMeta } from '@aws-amplify/graphql-docs-generator';
const CODEGEN_WARNING = 'this is an auto generated file. This will be overwritten';
const LINE_DELIMITOR = '\n';

type Language = 'javascript' | 'graphql' | 'typescript' | 'flow' | 'angular';

/**
 * Utility class to format the generated GraphQL statements based on frontend language type
 */
export class GraphQLStatementsFormatter {
  private language: Language;

  private lintOverrides: string[];

  private headerComments: string[];

  private opTypeName?: string;

  private typesPath: string | null;

  private includeTypeScriptTypes: boolean;

  constructor(language: Language, operation: string, typesPath?: string) {
    this.language = language || 'graphql';
    this.opTypeName = {
      queries: 'Query',
      mutations: 'Mutation',
      subscriptions: 'Subscription',
    }[operation];
    this.lintOverrides = [];
    this.headerComments = [];
    this.typesPath = typesPath ? typesPath.replace(/.ts/i, '') : null;
    this.includeTypeScriptTypes = !!(this.language === 'typescript' && this.opTypeName && this.typesPath);
  }

  get typeDefs() {
    if (!this.includeTypeScriptTypes) return '';
    return [
      `import * as APITypes from '${this.typesPath}';`,
      `type Generated${this.opTypeName}<InputType, OutputType> = string & {`,
      `  __generated${this.opTypeName}Input: InputType;`,
      `  __generated${this.opTypeName}Output: OutputType;`,
      `};`,
    ].join(LINE_DELIMITOR);
  }

  format(statements: Map<string, GraphQLWithMeta>): string {
    switch (this.language) {
      case 'javascript':
        this.headerComments.push(CODEGEN_WARNING);
        this.lintOverrides.push('/* eslint-disable */');
        return this.prettify(this.formatJS(statements));
      case 'typescript':
        this.headerComments.push(CODEGEN_WARNING);
        this.lintOverrides.push(...['/* tslint:disable */', '/* eslint-disable */']);
        return this.prettify(this.formatJS(statements));
      case 'flow':
        this.headerComments.push('@flow', CODEGEN_WARNING);
        return this.prettify(this.formatJS(statements));
      default:
        this.headerComments.push(CODEGEN_WARNING);
        return this.prettify(this.formatGraphQL(statements));
    }
  }

  formatGraphQL(statements: Map<string, GraphQLWithMeta>): string {
    const headerBuffer = this.headerComments.map(comment => `# ${comment}`).join(LINE_DELIMITOR);
    const statementsBuffer = statements ? [...statements.values()].map(s => s.graphql).join(LINE_DELIMITOR) : '';
    const formattedOutput = [headerBuffer, LINE_DELIMITOR, statementsBuffer].join(LINE_DELIMITOR);
    return formattedOutput;
  }

  formatJS(statements: Map<string, GraphQLWithMeta>): string {
    const lintOverridesBuffer = this.lintOverrides.join(LINE_DELIMITOR);
    const headerBuffer = this.headerComments.map(comment => `// ${comment}`).join(LINE_DELIMITOR);
    const formattedStatements = [];
    if (statements) {
      for (const [key, { graphql, operationName, operationType }] of statements) {
        const typeTag = this.buildTypeTag(operationName, operationType);
        const formattedGraphQL = prettier.format(graphql, { parser: 'graphql' });
        formattedStatements.push(`export const ${key} = /* GraphQL */ \`${formattedGraphQL}\`${typeTag}`);
      }
    }
    const typeDefs = this.includeTypeScriptTypes ? [LINE_DELIMITOR, this.typeDefs] : [];
    const formattedOutput = [lintOverridesBuffer, headerBuffer, ...typeDefs, LINE_DELIMITOR, ...formattedStatements].join(LINE_DELIMITOR);
    return formattedOutput;
  }

  buildTypeTag(operationName?: string, operationType?: string): string {
    if (!this.includeTypeScriptTypes || operationName === undefined || operationType === undefined) return '';

    const operationDef = { operationName, operationType };
    const resultTypeName = `APITypes.${interfaceNameFromOperation(operationDef)}`;
    const variablesTypeName = `APITypes.${interfaceVariablesNameFromOperation(operationDef)}`;

    return ` as Generated${this.opTypeName}<${variablesTypeName}, ${resultTypeName}>;`;
  }

  prettify(output: string): string {
    const parserMap: { [key in Language]: BuiltInParserName } = {
      javascript: 'babel',
      graphql: 'graphql',
      typescript: 'typescript',
      flow: 'flow',
      angular: 'graphql',
    };
    return prettier.format(output, { parser: parserMap[this.language || 'graphql'] });
  }
}
