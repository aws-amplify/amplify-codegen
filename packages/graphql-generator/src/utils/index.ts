export { GraphQLStatementsFormatter } from './GraphQLStatementsFormatter';
export { statementsTargetToFileExtensionMap } from './maps';
