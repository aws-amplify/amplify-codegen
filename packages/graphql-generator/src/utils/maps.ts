import { StatementsTarget, FileExtension } from '../typescript';

export const statementsTargetToFileExtensionMap: { [key in StatementsTarget]: FileExtension } = {
  javascript: 'js',
  graphql: 'graphql',
  flow: 'js',
  typescript: 'ts',
  angular: 'graphql',
};
