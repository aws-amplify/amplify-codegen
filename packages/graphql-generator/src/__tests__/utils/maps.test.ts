import { statementsTargetToFileExtensionMap } from '../../utils';

describe('maps', () => {
  test('statementsTargetToFileExtensionMap', () => {
    expect(statementsTargetToFileExtensionMap).toMatchSnapshot();
  });
});
