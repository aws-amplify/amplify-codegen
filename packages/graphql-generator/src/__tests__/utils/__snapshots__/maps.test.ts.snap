// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maps statementsTargetToFileExtensionMap 1`] = `
Object {
  "angular": "graphql",
  "flow": "js",
  "graphql": "graphql",
  "javascript": "js",
  "typescript": "ts",
}
`;
